### DISCUSSION TITLE
Windows provisioning includes the patch step 4 times

### DISCUSSION THREAD

**Discussion by Ruluk** [gusztavvargadr/packer#247] [General]
When inspecting the Packer templates, I noticed that one of the things that is done in all provisioning cases for Windows is updating.

That is obviously fine, but I noticed that this is done four times in series. It can be seen in the `src/w/packer/provisioners/chef/template.json` file. The following chef provisioning step is repeated 4 times:

```json
    {
      "type": "chef-solo",
      "guest_os_type": "{{user `chef_guest_os_type`}}",
      "skip_install": "{{user `chef_skip_install`}}",
      "staging_directory": "{{user `chef_staging_directory`}}",
      "remote_cookbook_paths": "{{user `chef_destination`}}/cookbooks",
      "run_list": "{{user `chef_run_list_patch`}}"
    },
```

Could you share with us more context on this? What happens if I remove those repeated steps?

**Comment by gusztavvargadr** [comment]
@Ruluk thanks for raising this.

On some Windows variants (especially on Server, and mainly the older versions, e.g. on 2016) it was common that Windows Update did not install all patches in one go, but in several rounds. First it patched the update system itself, then some basic patches like service packs, then additional updates on top of that. This is a very lazy solution to overcome this and ensure that all patches are applied on all variants.

I'm currently working on an approach where this is not needed, there will be a single provisioning step with Packer's built-in retry mechanism, reducing the number of reboots and the overall build time significantly.

I hope that helps, let me know if you have any other suggestions.
